File: frontend/src/lib/schema-detector.ts
/**
 * Client-side schema detector — TypeScript port of backend/app/services/schema_detector.py.
 *
 * Detects column mappings for 17 target fields using name hints + sample value scoring.
 * Used when USE_CLIENT_ETL=true so the browser does the parsing instead of server-side pandas.
 */

import Papa from "papaparse";
import {
  TARGET_COLUMNS,
  COLUMN_ALIASES,
  type TargetColumn,
  type UploadResult,
} from "@/types/upload";

// ---------------------------------------------------------------------------
// Text cleaning (mirrors Python _clean)
// ---------------------------------------------------------------------------
function clean(name: string): string {
  return name.toLowerCase().trim().replace(/[\s-]+/g, "_");
}

// ---------------------------------------------------------------------------
// Levenshtein similarity
// ---------------------------------------------------------------------------
function levenshteinSimilarity(a: string, b: string): number {
  if (!a.length || !b.length) return 1.0;
  const la = a.length;
  const lb = b.length;
  const d: number[][] = Array.from({ length: la + 1 }, () => Array(lb + 1).fill(0));
  for (let i = 0; i <= la; i++) d[i][0] = i;
  for (let j = 0; j <= lb; j++) d[0][j] = j;
  for (let i = 1; i <= la; i++) {
    for (let j = 1; j <= lb; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      d[i][j] = Math.min(d[i - 1][j] + 1, d[i][j - 1] + 1, d[i - 1][j - 1] + cost);
    }
  }
  return 1 - d[la][lb] / Math.max(la, lb);
}

// ---------------------------------------------------------------------------
// Name similarity — exact substring or fuzzy Levenshtein > 0.82
// ---------------------------------------------------------------------------
function nameSimilarity(colName: string, hints: string[]): number {
  const c = clean(colName);
  for (const hint of hints) {
    const h = clean(hint);
    if (c.includes(h) || h.includes(c)) return 1.0;
    if (levenshteinSimilarity(c, h) > 0.82) return 0.82;
  }
  return 0.0;
}

// ---------------------------------------------------------------------------
// Sample validators
// ---------------------------------------------------------------------------
const DATE_REGEX = [
  /^\d{4}-\d{2}-\d{2}/,          // 2024-01-15
  /^\d{2}\/\d{2}\/\d{4}/,       // 15/01/2024
  /^\d{2}-\d{2}-\d{4}/,         // 15-01-2024
  /^\d{2}-[A-Za-z]{3}-\d{4}/,   // 15-Jan-2024
  /^\d{1,2}\s+[A-Za-z]{3,9}\s+\d{4}/, // 15 January 2024
  /^\d{4}\/\d{2}\/\d{2}/,       // 2024/01/15
];

function isDateLike(value: unknown): boolean {
  if (value instanceof Date) return true;
  if (typeof value === "number" && value > 20000 && value < 60000) return true; // Excel serial date
  const s = String(value ?? "").trim();
  if (!s) return false;
  if (DATE_REGEX.some((rx) => rx.test(s))) return true;
  const t = Date.parse(s);
  return !isNaN(t);
}

function isPositiveNumeric(value: unknown): boolean {
  const s = String(value ?? "").trim();
  if (!s) return false;
  // Strip currency tokens (SAR, ر.س, ريال, etc.)
  const cleaned = s.replace(/SAR|S\.A\.R|ر\.س|ريال|﷼|,/gi, "").replace(/[^0-9.\-]/g, "").trim();
  if (!cleaned) return false;
  const n = parseFloat(cleaned);
  return !isNaN(n) && n >= 0;
}

function isText(value: unknown): boolean {
  return typeof value === "string" && value.trim().length > 1;
}

function isNonEmptyString(value: unknown): boolean {
  return typeof value === "string" && value.trim().length > 0;
}

// ---------------------------------------------------------------------------
// Field definitions — mirrors Python FIELD_PATTERNS
// ---------------------------------------------------------------------------
type SampleValidator = (value: unknown) => boolean;

interface FieldDef {
  target: TargetColumn;
  validator: SampleValidator;
  skipIf?: (colCleaned: string) => boolean;
}

const FIELD_DEFS: FieldDef[] = [
  { target: "expiry_date", validator: isDateLike,
    skipIf: (c) => !/(expir|exp_|best_before|bestbefore|صلاح|انتهاء)/.test(c) },
  { target: "transaction_at", validator: isDateLike },
  { target: "item_name", validator: isText },
  { target: "barcode", validator: (v) => String(v ?? "").trim().length >= 6 },
  { target: "item_sku", validator: (v) => String(v ?? "").trim().length > 1 },
  { target: "brand", validator: isText },
  { target: "pack_size", validator: isText },
  { target: "storage_type", validator: isText },
  { target: "current_stock", validator: isPositiveNumeric,
    skipIf: (c) => /^(qty|quantity|qnty|units|كمية|الكمية)$/.test(c) },
  { target: "quantity", validator: isPositiveNumeric },
  { target: "reorder_level", validator: isPositiveNumeric },
  { target: "max_stock", validator: isPositiveNumeric },
  { target: "unit_price", validator: isPositiveNumeric,
    skipIf: (c) => /(cost|purchase|buy|landed|شراء|تكلفة)/.test(c) },
  { target: "cost_price", validator: isPositiveNumeric },
  { target: "total_amount", validator: isPositiveNumeric },
  { target: "category_name", validator: isText },
  { target: "batch_number", validator: isNonEmptyString },
];

// ---------------------------------------------------------------------------
// Score a column against all target fields, return best match
// ---------------------------------------------------------------------------
function scoreColumn(
  colName: string,
  sampleValues: unknown[]
): { target: TargetColumn; score: number } | null {
  const c = clean(colName);
  let bestTarget: TargetColumn | null = null;
  let bestScore = 0;

  for (const def of FIELD_DEFS) {
    if (def.skipIf?.(c)) continue;

    const hints = COLUMN_ALIASES[def.target] || [];
    const nameScore = nameSimilarity(colName, hints);

    const samples = sampleValues.slice(0, 20).filter((v) => v != null && v !== "");
    const matchCount = samples.filter((v) => def.validator(v)).length;
    const sampleScore = samples.length > 0 ? matchCount / samples.length : 0;

    const combined = nameScore * 0.7 + sampleScore * 0.3;
    if (combined > bestScore) {
      bestScore = combined;
      bestTarget = def.target;
    }
  }

  if (bestTarget && bestScore >= 0.4) {
    return { target: bestTarget as TargetColumn, score: bestScore };
  }
  return null;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Parse a CSV file with PapaParse and detect column mappings.
 * Returns the same shape as the server-side UploadResult so the ColumnMapper UI works unchanged.
 */
export function detectSchemaFromCSV(file: File): Promise<Omit<UploadResult, "upload_id">> {
  return new Promise((resolve, reject) => {
    Papa.parse(file, {
      header: true,
      skipEmptyLines: true,
      preview: 100, // Only read first 100 rows for detection
      complete(results) {
        const rows = results.data as Record<string, unknown>[];
        const allColumns = results.meta.fields || [];

        // Count total rows (re-parse without preview for accurate count)
        Papa.parse(file, {
          header: false,
          skipEmptyLines: true,
          complete(countResults) {
            const totalRows = (countResults.data as unknown[][]).length;
            const detection = detectFromRows(allColumns, rows, totalRows, file.name);
            resolve(detection);
          },
          error(err) {
            reject(err);
          },
        });
      },
      error(err) {
        reject(err);
      },
    });
  });
}

/**
 * Detect schema from pre-parsed rows (e.g., already parsed by PapaParse in the upload page).
 */
export function detectFromRows(
  columns: string[],
  sampleRows: Record<string, unknown>[],
  totalRows: number,
  filename: string = "upload.csv"
): Omit<UploadResult, "upload_id"> {
  const detectedColumns: Record<string, string> = {};
  const confidenceScores: Record<string, number> = {};
  const usedTargets = new Set<string>();

  for (const col of columns) {
    const sampleValues = sampleRows.map((r) => r[col]);
    const result = scoreColumn(col, sampleValues);

    if (result && !usedTargets.has(result.target)) {
      detectedColumns[col] = result.target;
      confidenceScores[col] = result.score;
      usedTargets.add(result.target);
    }
  }

  const mappedTargets = Object.values(detectedColumns);
  const unmappedColumns = columns.filter((c) => !detectedColumns[c]);

  // Determine file kind
  const hasTransactionAt = mappedTargets.includes("transaction_at");
  let suggestedFileKind: string = hasTransactionAt ? "sales_history" : "inventory_snapshot";

  // If no sales date detected, remap quantity → current_stock (Python behavior)
  if (!hasTransactionAt) {
    for (const [src, tgt] of Object.entries(detectedColumns)) {
      if (tgt === "quantity") {
        detectedColumns[src] = "current_stock";
        break;
      }
    }
  }

  return {
    filename,
    file_size: 0, // Will be set by caller
    mime_type: "text/csv",
    row_count: totalRows,
    detected_columns: detectedColumns,
    confidence_scores: confidenceScores,
    unmapped_columns: unmappedColumns,
    sample_rows: sampleRows.slice(0, 5),
    suggested_file_kind: suggestedFileKind,
    schema_valid: Object.keys(detectedColumns).length > 0,
  };
}

/**
 * Detect Foodics/Salla export signature.
 * Returns a suggested file kind if the column set matches a known POS export.
 */
export function detectPOSSignature(columns: string[]): string | null {
  const c = columns.map(clean);

  const foodicsMarkers = ["date", "item", "qty", "amount", "cost"];
  const foodicsScore = foodicsMarkers.filter((m) => c.some((col) => col.includes(m))).length;
  if (foodicsScore >= 4) return "foodics_export";

  const sallaMarkers = ["order", "product", "quantity", "price", "total"];
  const sallaScore = sallaMarkers.filter((m) => c.some((col) => col.includes(m))).length;
  if (sallaScore >= 4) return "salla_export";

  return null;
}

File: backend/app/services/schema_detector.py
from datetime import datetime
from typing import Dict, List, Tuple
import re

import pandas as pd


def is_date_like(value) -> bool:
    if pd.isna(value):
        return False
    if isinstance(value, (datetime, pd.Timestamp)):
        return True
    if isinstance(value, (int, float)) and value > 20000:
        return True
    if isinstance(value, str):
        date_patterns = [
            r"\d{4}-\d{2}-\d{2}",
            r"\d{2}/\d{2}/\d{4}",
            r"\d{2}-\d{2}-\d{4}",
            r"\d{2}-[A-Za-z]{3}-\d{4}",
            r"\d{1,2}\s+[A-Za-z]{3,9}\s+\d{4}",
            r"\d{4}/\d{2}/\d{2}",
        ]
        return any(re.search(pattern, value) for pattern in date_patterns)
    return False


def is_positive_numeric(value) -> bool:
    if pd.isna(value):
        return False
    try:
        cleaned = re.sub(r"(SAR|S\.A\.R|ر\.س|ريال|﷼|,)", "", str(value), flags=re.IGNORECASE)
        cleaned = re.sub(r"[^0-9.\-]", "", cleaned)
        return float(cleaned) >= 0
    except (ValueError, TypeError):
        return False


def is_text(value) -> bool:
    return isinstance(value, str) and len(str(value).strip()) > 1


FIELD_PATTERNS = {
    "expiry_date": {
        "name_hints": [
            "expiry", "expiration", "expires", "best_before", "bestbefore", "use_by", "exp_date",
            "تاريخ_الصلاحية", "صلاحية", "انتهاء", "تاريخ الانتهاء",
        ],
        "sample_validator": is_date_like,
    },
    "transaction_at": {
        "name_hints": [
            "date", "time", "datetime", "transaction_date", "sale_date", "sales_date",
            "billing_date", "invoice_date", "bill_date", "dt", "day", "sold_at",
            "التاريخ", "تاريخ", "وقت", "تاريخ البيع",
        ],
        "sample_validator": is_date_like,
    },
    "item_name": {
        "name_hints": [
            "item", "product", "name", "description", "item_name", "product_name", "particulars",
            "goods", "article", "items", "material", "اسم", "اسم الصنف", "اسم المنتج", "منتج", "صنف", "المادة",
        ],
        "sample_validator": is_text,
    },
    "barcode": {
        "name_hints": ["barcode", "bar_code", "ean", "upc", "gtin", "باركود"],
        "sample_validator": lambda x: len(str(x).strip()) >= 6,
    },
    "item_sku": {
        "name_hints": [
            "sku", "item_sku", "product_sku", "code", "item_code", "product_code", "article_code",
            "part_no", "item_no", "كود", "رمز", "رمز الصنف",
        ],
        "sample_validator": lambda x: len(str(x).strip()) > 1,
    },
    "brand": {
        "name_hints": ["brand", "make", "manufacturer", "ماركة", "علامة", "الشركة"],
        "sample_validator": is_text,
    },
    "pack_size": {
        "name_hints": ["pack", "pack_size", "size", "uom", "unit_size", "case_pack", "عبوة", "حجم", "وحدة"],
        "sample_validator": is_text,
    },
    "storage_type": {
        "name_hints": ["storage", "storage_type", "temperature", "ambient", "chilled", "frozen", "تخزين", "حرارة"],
        "sample_validator": is_text,
    },
    "current_stock": {
        "name_hints": [
            "current_stock", "stock", "stock_qty", "current_qty", "available", "available_qty",
            "on_hand", "onhand", "balance", "inventory", "closing_stock", "رصيد", "مخزون", "الكمية المتوفرة",
        ],
        "sample_validator": is_positive_numeric,
    },
    "quantity": {
        "name_hints": ["qty", "quantity", "sold_qty", "sales_qty", "units", "pieces", "pcs", "nos", "qnty", "كمية", "الكمية"],
        "sample_validator": is_positive_numeric,
    },
    "reorder_level": {
        "name_hints": ["reorder", "reorder_level", "reorder_point", "min_stock", "minimum_stock", "min", "حد الطلب", "الحد الادنى"],
        "sample_validator": is_positive_numeric,
    },
    "max_stock": {
        "name_hints": ["max_stock", "maximum_stock", "max", "par_level", "capacity", "الحد الاعلى"],
        "sample_validator": is_positive_numeric,
    },
    "unit_price": {
        "name_hints": [
            "price", "rate", "unit_price", "mrp", "selling_price", "sale_price", "retail_price",
            "sp", "unit_rate", "per_unit", "selling", "سعر", "سعر البيع",
        ],
        "sample_validator": is_positive_numeric,
    },
    "cost_price": {
        "name_hints": ["cost", "cost_price", "purchase_price", "cp", "buy_price", "landed_cost", "purchase_rate", "تكلفة", "سعر الشراء"],
        "sample_validator": is_positive_numeric,
    },
    "total_amount": {
        "name_hints": ["total", "amount", "net_amount", "gross_amount", "total_amount", "bill_amount", "net", "value", "revenue", "sale_value", "amt", "اجمالي", "المبلغ"],
        "sample_validator": is_positive_numeric,
    },
    "category_name": {
        "name_hints": ["category", "cat", "department", "dept", "group", "section", "type", "class", "division", "grp", "تصنيف", "قسم", "فئة"],
        "sample_validator": is_text,
    },
    "batch_number": {
        "name_hints": ["batch", "batch_no", "batch_number", "lot", "lot_no", "lot_number", "تشغيلة", "دفعة"],
        "sample_validator": lambda x: len(str(x).strip()) > 0,
    },
}


class SchemaDetector:
    # ── POS export signatures ──────────────────────────────────────
    FOODICS_MARKERS = {"date", "item", "qty", "amount", "cost"}
    SALLA_MARKERS = {"order", "product", "quantity", "price", "total"}

    @classmethod
    def detect_pos_signature(cls, columns: list) -> str | None:
        """Detect Foodics/Salla export by column name signature."""
        cleaned = [c.lower().strip().replace(" ", "_").replace("-", "_") for c in columns]
        flat = set()
        for c in cleaned:
            for word in c.split("_"):
                flat.add(word)

        foodics_score = sum(1 for m in cls.FOODICS_MARKERS if m in flat)
        if foodics_score >= 4:
            return "foodics_export"

        salla_score = sum(1 for m in cls.SALLA_MARKERS if m in flat)
        if salla_score >= 4:
            return "salla_export"

        return None

    def detect(self, df: pd.DataFrame) -> Dict:
        result = {}
        confidence = {}
        used_targets = set()

        # Check for known POS export signatures first
        pos_signature = self.detect_pos_signature(list(df.columns))
        if pos_signature:
            # For known POS exports, boost confidence for matching columns
            pass

        for col in df.columns:
            best_field, best_score = self._score_column(col, df[col], used_targets)
            if best_field and best_score >= 0.4:
                result[col] = best_field
                confidence[col] = best_score
                used_targets.add(best_field)

        # If the file has no sale date, merchant "quantity" columns are usually stock-on-hand.
        has_sales_date = "transaction_at" in result.values()
        if not has_sales_date:
            for col, target in list(result.items()):
                if target == "quantity":
                    result[col] = "current_stock"

        unmapped = [c for c in df.columns if c not in result]

        return {
            "detected_columns": result,
            "confidence_scores": confidence,
            "unmapped_columns": unmapped,
            "sample_rows": df.head(5).fillna("").to_dict(orient="records"),
            "total_rows": len(df),
            "suggested_file_kind": pos_signature or ("sales_history" if has_sales_date else "inventory_snapshot"),
        }


    def detect_columns(self, df: pd.DataFrame) -> Dict[str, str]:
        """Backward-compatible detector API used by older tests/tools."""
        legacy_result: Dict[str, str] = {}
        used = set()
        for col in df.columns:
            c = self._clean(col)
            target = None
            if any(x in c for x in ["category", "cat", "department", "dept", "item_type", "product_type", "type"]):
                target = "category"
            elif any(x in c for x in ["product_name", "item_name", "description", "product", "item", "name"]):
                target = "product_name"
            elif any(x in c for x in ["qty", "quantity", "sold_qty", "sales_qty", "units"]):
                target = "quantity"
            elif any(x in c for x in ["price", "rate", "selling_price", "sale_price", "unit_price"]):
                target = "unit_price"
            elif any(x in c for x in ["sku", "code"]):
                target = "sku"
            elif any(x in c for x in ["stock", "on_hand", "available", "inventory"]):
                target = "current_stock"
            if target and target not in used:
                legacy_result[col] = target
                used.add(target)
        if legacy_result:
            return legacy_result
        detected = self.detect(df)["detected_columns"]
        legacy = {"item_name": "product_name", "item_sku": "sku", "category_name": "category", "sell_price": "unit_price"}
        return {source: legacy.get(target, target) for source, target in detected.items()}

    def get_confidence_scores(self, df: pd.DataFrame) -> Dict[str, float]:
        """Backward-compatible confidence API: source column -> score."""
        return {source: 1.0 for source in self.detect_columns(df).keys()}

    def _score_column(self, col_name: str, series: pd.Series, used: set) -> Tuple[str | None, float]:
        best_field = None
        best_score = 0.0
        col_lower = self._clean(col_name)
        samples = series.dropna().head(20).tolist()

        for field_name, field_info in FIELD_PATTERNS.items():
            if field_name in used:
                continue

            # A generic "date" column in sales exports should not become expiry_date.
            if field_name == "expiry_date" and not any(token in col_lower for token in ["exp", "expiry", "expiration", "best_before", "bestbefore", "صلاح", "انتهاء"]):
                continue

            # Generic qty in sales files should map to quantity, not current stock. Stock exports usually say stock/balance/on hand.
            if field_name == "current_stock" and col_lower in {"qty", "quantity", "qnty", "units", "كمية", "الكمية"}:
                continue

            # Purchase/cost columns often include Arabic "سعر" too; do not let unit_price steal them.
            if field_name == "unit_price" and any(token in col_lower for token in ["cost", "purchase", "buy", "landed", "شراء", "تكلفة"]):
                continue

            name_score = self._name_similarity(col_lower, field_info["name_hints"])
            sample_score = self._sample_match(samples, field_info["sample_validator"])
            combined = (name_score * 0.7) + (sample_score * 0.3)

            if combined > best_score:
                best_score = combined
                best_field = field_name

        return best_field, best_score

    def _clean(self, col_name: str) -> str:
        return str(col_name).lower().strip().replace(" ", "_").replace("-", "_")

    def _name_similarity(self, col_name: str, hints: List[str]) -> float:
        col_lower = col_name.lower()
        for hint in hints:
            hint_lower = self._clean(hint)
            if hint_lower in col_lower or col_lower in hint_lower:
                return 1.0
            if self._levenshtein_similarity(col_lower, hint_lower) > 0.82:
                return 0.82
        return 0.0

    def _sample_match(self, samples: List, validator) -> float:
        if not samples:
            return 0.0
        matches = sum(1 for s in samples if validator(s))
        return matches / len(samples)

    def _levenshtein_similarity(self, s1: str, s2: str) -> float:
        if len(s1) < len(s2):
            return self._levenshtein_similarity(s2, s1)
        if len(s2) == 0:
            return 0.0
        previous_row = range(len(s2) + 1)
        for i, c1 in enumerate(s1):
            current_row = [i + 1]
            for j, c2 in enumerate(s2):
                insertions = previous_row[j + 1] + 1
                deletions = current_row[j] + 1
                substitutions = previous_row[j] + (c1 != c2)
                current_row.append(min(insertions, deletions, substitutions))
            previous_row = current_row
        distance = previous_row[-1]
        max_len = max(len(s1), len(s2))
        return 1 - (distance / max_len) if max_len > 0 else 1.0
